### PR TITLE
JS: add sanitizer for relative ".." in js/path-injection

### PR DESCRIPTION
**Pull request by erik-krogh**
This PR adds the following sanitizer in `js/path-injection`: 

```JavaScript 
var relative = path.relative(webroot, pathname);
if(relative.startsWith(".." + path.sep) || relative == "..") {
    return res.sendStatus(404); // pathname is bad
} else {
   fs.writeFileSync(pathname, contents); // pathname is sanitized
}
```

This sanitizer does not work with Windows drive letters (a relative path from `C:\foo\bar` to `D:\baz\bla` is just `D:\baz\bla`). But I still think it is good enough for us to use it as a sanitizer. 

The sanitizer is used in the fix for CVE-2018-3787, and adding the sanitizer fixes all the new FPs that arose as part of https://github.com/Semmle/ql/pull/2810. 

An [evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-asger.northeurope.cloudapp.azure.com_1581656739179) look OK performance wise .

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll [hunk 1]
```diff
@@ -355,6 +353,35 @@ module TaintedPath {
...
+  class RelativePathContainsDotDotGuard extends DataFlow::BarrierGuardNode {
+    StringOps::StartsWith startsWith;
+    DataFlow::CallNode relativeCall;
+
+    RelativePathContainsDotDotGuard() {
+      this = startsWith and
+      relativeCall = DataFlow::moduleImport("path").getAMemberCall("relative") and
+      startsWith.getBaseString().getALocalSource() = relativeCall and
+      exists(DataFlow::Node subString | subString = startsWith.getSubstring() |
+        subString.mayHaveStringValue("..")
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll]: Here and below. Lets support the `"../"` prefixes as well.

File: javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll [hunk 2]
```diff
@@ -12,19 +12,17 @@ module TaintedPath {
...
-      result instanceof Label::PosixPath
-    }
+    DataFlow::FlowLabel getAFlowLabel() { result instanceof Label::PosixPath }
   }
 
   /**
    * A data flow sink for tainted-path vulnerabilities.
    */
   abstract class Sink extends DataFlow::Node {
+    Sink() { not this instanceof Sanitizer }
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll]: This seems out of place for this PR, and if we want it, then it is perhaps something that should be implemented in `Configuration.qll` as something like `isReallySink(Node sink) { isSink(sink) and not isBarrier(sink) }`

File: javascript/ql/test/query-tests/Security/CWE-022/TaintedPath/normalizedPaths.js [hunk 3]
```diff
@@ -249,3 +249,26 @@ app.get('/resolve-path', (req, res) => {
...
+	
+  var relative = pathModule.relative(self.webroot, path);
+  if(relative.startsWith(".." + pathModule.sep) || relative == "..") {
+    fs.readFileSync(path); // NOT OK! 
+  } else {
+    fs.readFileSync(path); // OK! 
+  }
+
+  let newpath = pathModule.normalize(p);
+  var relativePath = path.relative(path.normalize(workspaceDir), newpath);
```
Comment (esbena) [inline javascript/ql/test/query-tests/Security/CWE-022/TaintedPath/normalizedPaths.js]: In QL, do we want to support the pattern of `path.normalize(path.relative(x, newpath));`? Programmers may want to be *extra* safe that they are doing the right thing, so the redundancy is not completely unrealistic IMO.

File: javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll (hunk 2 above)
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll]: It was part of my testing for the PR and it should have been removed, but I forgot.

File: javascript/ql/test/query-tests/Security/CWE-022/TaintedPath/normalizedPaths.js (hunk 3 above)
Comment (erik-krogh) [reply javascript/ql/test/query-tests/Security/CWE-022/TaintedPath/normalizedPaths.js]: I didn't find anyone using that pattern, but it doesn't sound unreasonable, so I added support for it.

File: javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll [hunk 4]
```diff
@@ -355,6 +351,47 @@ module TaintedPath {
...
+  /**
+   * A sanitizer that recognizes the following pattern:
+   *   var relative = path.relative(webroot, pathname);
+   *   if(relative.startsWith(".." + path.sep) || relative == "..") {
+   *     // pathname is unsafe
+   *   } else {
+   *     // pathname is safe
+   *   }
+   */
+  class RelativePathContainsDotDotGuard extends DataFlow::BarrierGuardNode {
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll]: `RelativePathContainsDotDotGuard ` -> `RelativePathStartsWithDotDotSanitizer`

The other guards are called `Sanitizer` so we might as well be consistent, and this handles `StartsWith` checks, not inclusion checks.

**Comment by asgerf** [comment]
@esbena any more comments? (you're still requesting changes)

**Pull request merged**